### PR TITLE
move regex related funcs to its own regex.js file

### PR DESCRIPTION
**Pull request by kmachado**
move regex related funcs to its own regex.js file
move regex related tests to regex.spec.js

### CODE REVIEW

File: src/regex.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  const postingsByWord = (result, word) => {
+    result.push(postings(RegExp(word, 'gi'), str))
+    return result
+  }
+  return _.reduce(postingsByWord, [])(_.words(string))
+}
+
+export const insertAtIndex = (index, val, str) => str.slice(0, index) + val + str.slice(index)
+
+export const mergeRanges = (p1, p2) => ((p2[0] <= p1[1]) && [[p1[0], _.max([p1[1], p2[1]])]]) || [p1, p2]
```
Comment (daedalus28) [inline src/regex.js]: `]])]])` o_O

File: src/regex.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  }
+  return result
+}
+
+export const postingsForWords = (string, str) => {
+  const postingsByWord = (result, word) => {
+    result.push(postings(RegExp(word, 'gi'), str))
+    return result
+  }
+  return _.reduce(postingsByWord, [])(_.words(string))
```
Comment (daedalus28) [inline src/regex.js]: this could probably be a flow

File: README.md [hunk 3]
```diff
@@ -142,11 +142,44 @@ Example: `{ a: { b: { c: 1 } } } => { 'a.b.c' : 1 }`.
...
+### highlight
+`(start, end, postings, string) -> string`
+takes start and end tags to enclose the postings into the string
+
+### flattenPostings
+`flattenPostings -> [] -> []`
+takes an array of postings or postingsForWords and return a flattened and ordered array of postings
+
+### mergeRanges
+`mergeRangesr -> (r1, r2) -> []`
```
Comment (daedalus28) [inline README.md]: arrrrrrrrrr

File: README.md [hunk 4]
```diff
@@ -142,11 +142,44 @@ Example: `{ a: { b: { c: 1 } } } => { 'a.b.c' : 1 }`.
...
+## Regex
 
 ### testRegex
 `regex -> string -> bool` Just like rambda test, creates a function to test a regex on a string.
 
+### makeRegex
+`makeRegex -> options -> string -> regex` Just a curried implementation of the RegExp construction.
+
+### makeAndTest
+`makeAndTest -> options -> (string) -> ` makes and tests a RegExp with makeRegex and testRegex
```
Comment (daedalus28) [inline README.md]: options:{} -> string -> (string -> bool)

File: src/regex.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+export const insertAtIndex = (index, val, str) => str.slice(0, index) + val + str.slice(index)
```
Comment (daedalus28) [inline src/regex.js]: move to array.js

File: src/regex.js (hunk 1 above)
Comment (daedalus28) [inline src/regex.js]: move to array.js

File: src/regex.js (hunk 2 above)
Comment (daedalus28) [inline src/regex.js]: dont expose

File: src/regex.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  return result
+}
+
+export const flattenPostings = arrOfPostingsByWord => {
+  const sortedFlatten = _.flow(
+    _.flatten,
+    _.sortBy([0, 1])
+  )(arrOfPostingsByWord)
+  return _.reduce(mergeRangesOver, [])(sortedFlatten)
+}
```
Comment (daedalus28) [inline src/regex.js]: dont expose

Comment (daedalus28) [inline src/regex.js]:
```js
export const flattenPostings = _.flow(
  _.flatten,
  _.sortBy([0, 1]),
  _.reduce(mergeRangesOver, [])
)
```

File: src/regex.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+let push = (val, arr) => arr.concat([val])
+export const postingsForWords = (string, str) => _.reduce(
+  (result, word) => push(postings(RegExp(word, 'gi'), str), result), []
+)(_.words(string))
+
+export const insertAtIndex = (index, val, str) => str.slice(0, index) + val + str.slice(index)
+
+export const mergeRanges = (p1, p2) => ((p2[0] <= p1[1]) && [[p1[0], _.max([p1[1], p2[1]])]]) || [p1, p2]
```
Comment (daedalus28) [inline src/regex.js]: add some spacing

File: src/regex.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+const mergeRangesOver = (result, posting) => {
```
Comment (daedalus28) [inline src/regex.js]: what we really want is just a version of mergeRanges that can take more than two ranges

Comment (daedalus28) [inline src/regex.js]: e.g. `mergeAllRanges`

File: src/regex.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  let result = []
+
+  while (match) {
+    result.push([match.index, regex.lastIndex])
+    match = regex.exec(str)
+  }
+  return result
+}
+
+let push = (val, arr) => arr.concat([val])
```
Comment (daedalus28) [inline src/regex.js]: move to array

File: src/regex.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+export const postingsForWords = (string, str) => _.reduce(
+  (result, word) => push(postings(RegExp(word, 'gi'), str), result), []
+)(_.words(string))
+
+export const insertAtIndex = (index, val, str) => str.slice(0, index) + val + str.slice(index)
```
Comment (daedalus28) [inline src/regex.js]: move to array

File: src/regex.js (hunk 9 above)
Comment (daedalus28) [inline src/regex.js]: also add to docs

File: src/regex.js (hunk 7 above)
Comment (daedalus28) [inline src/regex.js]: move to array

**Comment by kmachado** [comment]
ready to go...

**Pull request merged**